include/boost/beast/core/buffers_generator.hpp

```diff
@@ -0,0 +1,229 @@
+//
+// Copyright (c) 2020 Vinnie Falco (vinnie dot falco at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/beast
+//
+
+#ifndef BOOST_BEAST_CORE_BUFFERS_GENERATOR_HPP
+#define BOOST_BEAST_CORE_BUFFERS_GENERATOR_HPP
+
+#include <boost/beast/core/detail/config.hpp>
+
+#include <boost/asio/write.hpp>
+#include <boost/asio/async_result.hpp>
+#include <boost/asio/compose.hpp>
+#include <boost/asio/coroutine.hpp>
+
+#include <boost/beast/core/buffer_traits.hpp>
+#include <boost/beast/core/stream_traits.hpp>
+#include <boost/beast/core/detail/type_traits.hpp>
+
+#include <boost/throw_exception.hpp>
+#include <type_traits>
+
+namespace boost {
+namespace beast {
+
+/** Determine if type satisfies the <em>BuffersGenerator</em> requirements.
+
+    This metafunction is used to determine if the specified type meets the
+    requirements for a buffers generator.
+
+    The static member `value` will evaluate to `true` if so, `false` otherwise.
+
+    @tparam T a type to check
+*/
+#ifdef BOOST_BEAST_DOXYGEN
+template <class T>
+struct is_buffers_generator
+    : integral_constant<bool, automatically_determined>
+{
+};
+#else
+template<class T, class = void>
+struct is_buffers_generator
+    : std::false_type
+{
+};
+
+template<class T>
+struct is_buffers_generator<
+    T, detail::void_t<decltype(
+        typename T::const_buffers_type(
+            std::declval<T&>().prepare(
+                std::declval<error_code&>())),
+        std::declval<T&>().consume(
+            std::size_t{})
+    )>> : std::true_type
+{
+};
+#endif
+
+//----------------------------------------------------------
+
+namespace detail {
+
+template <
+    class AsyncWriteStream,
+    class BuffersGenerator>
+struct write_buffers_generator_op
+    : boost::asio::coroutine
+{
+    write_buffers_generator_op(AsyncWriteStream& stream,
+                               BuffersGenerator generator)
```

Comment:
derp formatting

Reply:
```c++
    write_buffers_generator_op(
        AsyncWriteStream& s, BuffersGenerator g)
        : s_(s), g_(std::move(g))
    {
    }
```

Reply:
This is better but this is better yet:
``
    write_buffers_generator_op(
        AsyncWriteStream& s, BuffersGenerator g)
        : s_(s)
        , g_(std::move(g))
    {
    }
```

For two reasons, because it is easier to get a breakpoint, and because coverage will consider the two ctor-initializers to be separate lines.